complete rebuild

Ethan-Arrowood opened the pull request:
Ref #8 
Pretty much a new api with this PR. My brain needs a break but tomorrow I'll update tests as well as the readme.

delvedor commented on jwt.js:
```diff
@@ -1,61 +1,106 @@
 const fp = require('fastify-plugin')
 const JWT = require('jsonwebtoken')
 const assert = require('assert')
+const waterfall = require('async/waterfall')
```
Can you use [`steed`](https://github.com/mcollina/steed) instead of `async`?

Ethan-Arrowood replied:
I love how you guys have a module replacement for literally everything and they are also super fast 😄

delvedor commented on jwt.js:
```diff
@@ -1,61 +1,106 @@
+const waterfall = require('async/waterfall')
+
+function isFunction(object) {
+  return Object.prototype.toString.call(object) === '[object Function]'
```
Why not `typeof object === 'function'`?

Ethan-Arrowood commented:
Okay so I wrote some unit tests. Coverage isn't 100% but this is at least a starting point. I also moved some chunks of code around in the api itself. I'm puzzled on how to make this synchronous (if it even can be anymore).

Ethan-Arrowood commented:
Okay unit tests cover 100% across the table. Let me know what you think! README update will come soon.

Ethan-Arrowood commented:
Looks like there is a travis issue. I think it needs a manual restart.

mcollina commented on jwt.js:
```diff
@@ -1,61 +1,114 @@
 const fp = require('fastify-plugin')
 const JWT = require('jsonwebtoken')
-const assert = require('assert')
+const steed = require('steed')()
```
you should instantiate steed within the plugin itself.

mcollina commented on jwt.js:
```diff
@@ -1,61 +1,114 @@
+          } else {
+            callback(null, token)
+          }
+        })
```
why is this callback needed? can't you just do `JWT.sign(request.body, secret, rest, callback)`.

mcollina commented on jwt.js:
```diff
@@ -1,61 +1,114 @@
+    let {secret, resultProperty, ...rest} = options
+    steed.waterfall([
+      function getSecret (callback) {
+        secretCallback({}, {}, callback)
```
why are passing empty options here?

Ethan-Arrowood replied:
Hmm I believe the `sign` method only requires the `secret` and the `payload` (which comes from `request.body`). Ahh this is my mistake thought because in the RS256 instance where the `secret` is dynamic, this method still needs to pass the `request.header.authorization` token. Okay I'll have to update this in the API!

mcollina commented on jwt.js:
```diff
@@ -1,61 +1,114 @@
 }
 
-module.exports = fp(fastifyJWT, '>=0.13.1')
+module.exports = fp(fastifyJwt, '0.x')
```
can you make this `>= 0.39`.

Ethan-Arrowood replied:
I think updating `fastify` from 0.13 to 0.39 broke the unit tests.

This ran and passed in 0.13, but now fails stating its unfinished.
```js
test('fastify-jwt should expose jwt methods', t => {
  t.plan(2)
  const fastify = Fastify()
  fastify
    .register(jwt, { secret: 'supersecret' })
    .after(() => {
      t.ok(fastify.jwt.sign)
      t.ok(fastify.jwt.verify)
    })
})
```
I think the `.after` was removed. Was it replaced with something else? 

I tried doing this but it says `can't read property sign of undefined`
```js
test('fastify-jwt should expose jwt methods', t => {
  t.plan(2)
  const fastify = Fastify()
  fastify
    .register(jwt, { secret: 'supersecret' })
  t.ok(fastify.jwt.sign)
  t.ok(fastify.jwt.verify)
})
```

Ethan-Arrowood replied:
I'm quite puzzled on how to test this now. I'm not sure why but after I register `jwt` the fastify instance still returns undefined for `fastify.jwt`.

mcollina replied:
this would be async. place the tests within `.after()`.

Ethan-Arrowood replied:
<img width="1315" alt="screen shot 2018-01-17 at 11 38 31 am" src="https://user-images.githubusercontent.com/16144158/35054622-fc86ab9e-fb7a-11e7-83a6-85915fb2c496.png">
😕

mcollina replied:
we have removed the callback in register. Do not pass `t.error`.

Ethan-Arrowood replied:
<img width="1221" alt="screen shot 2018-01-17 at 12 48 25 pm" src="https://user-images.githubusercontent.com/16144158/35058050-bd0ec000-fb84-11e7-9445-f599fce2f89c.png">

Still an issue. I'll commit what I have as well as the broken test file in case you want to test it for yourself.

mcollina commented on jwt.js:
```diff
@@ -1,61 +1,114 @@
+  if (typeof secretCallback !== 'function') { secretCallback = wrapStaticSecretInCallback(secretCallback) }
+
+  let _requestProperty = options.userProperty || options.requestProperty || 'user'
+  let _resultProperty = options.resultProperty
```
I would remove those. Why are they needed? I would not overwrite anything.

Ethan-Arrowood replied:
See my more in depth answer below. TL;DR this is my first middleware and I abstracted this from `express-jwt`. If this is bad practice then I'll remove it.

mcollina commented on jwt.js:
```diff
@@ -1,61 +1,114 @@
+
+  let _requestProperty = options.userProperty || options.requestProperty || 'user'
+  let _resultProperty = options.resultProperty
+  const credentialsRequired = typeof options.credentialsRequired === 'undefined' ? true : options.credentialsRequired
```
Why was this option introduced? Could you remove it?

Ethan-Arrowood replied:
This is another thing I abstracted from `express-jwt`. After reviewing `node-jsonwebtoken` it seems like this is a not an actual option. I'll remove it and update api.

mcollina commented on jwt.js:
```diff
@@ -1,61 +1,114 @@
-    options = options || {}
-    return JWT.decode(token, options)
-  }
+    let decodedToken = JWT.decode(token, { complete: true }) || {}
```
what does this  `complete: true` option do? Can you add a comment explaining? Why are you defaulting this wit `|| {}`.

Ethan-Arrowood replied:
https://github.com/auth0/node-jsonwebtoken#jwtdecodetoken--options

i'll make sure to comment this and mention it in the readme.

mcollina commented on package.json:
```diff
@@ -4,7 +4,8 @@
   "scripts": {
-    "test": "standard && tap test.js"
+    "test": "standard && tap test.js",
+    "fix": "standard --fix"
```
can you remove this?

mcollina commented on jwt.js:
```diff
@@ -1,61 +1,114 @@
+        reply[_resultProperty] = result
+      } else {
+        request[_requestProperty] = result
+      }
```
I'm 👎  in doing this. Why do you need to modify these?

Ethan-Arrowood replied:
This was abstracted from `express-jwt`. If you needed access to the result of either sign or verify on your request/reply then this lets you specify where you can find it.
```js
fastify.register(jwt, {
  secret: 'shh',
  resultProperty: 'foo'
})
fastify.get(
  '/api',
  {
    beforeHandler: [
      function (req, res, done) {
        fastify.jwt.verify(req, res, done)
      },
      function (req, res, done) {
        console.log(res.foo) // <- the result of jwt.verify
      }
    ]
  }, 
  (request, reply) => reply.send({ fuzz: 'buzz' })
)
```

This is the first time I've ever really handled middleware so you tell me if this is bad practice or not.

mcollina replied:
I think that's bad practice. If we need to add things in this way we should do:

```js
request.verifyJwt(cb) // request.verifyJwt().then()
```

I'm starting to think that this PR should add new decorators to `Request`  and `Reply` and __keep and use__ the sign/verify methods that were attached to the main instance.

Ethan-Arrowood replied:
Interesting, how would this be implemented in a beforeHandler method? Or would we execute these functions directly in the route handler?

mcollina commented on jwt.js:
```diff
@@ -1,61 +1,114 @@
-'use strict'
```
This __must__ be restored.

delvedor commented:
Can you also explain why wrap the secret inside a callback?

Ethan-Arrowood commented:
Looks like I got some work to do! This file is heavily influenced by the `express-jwt` file. 

I can definitely add back `jwt.decode`

Wrapping the secret inside a callback helps with execution because no matter if the secret is just a simple string, or a sync function, or a asynchronous function it will work with the rest of the api. 

I originally tried implementing a ternary operation where it would check if secret is a string and if so just return it, otherwise execute the function and return but I was having issues handling the asynchronous case.

Ethan-Arrowood commented:
I have resolved a majority of the reviews you guys have proposed. The unit test file is completely broken now. I'm going to see if I can fix it and get it running again and then I'll push my work.

Ethan-Arrowood commented:
```js
    steed.waterfall([
      function getToken (callback) {
        parseToken(request, callback)
      },
      function decodeToken (token, callback) {
        decode(token, callback)
      },
      function getSecret (decodedToken, callback) {
        secretCallback(request, decodedToken, callback)
      },
      function verify (secret, callback) {
        JWT.verify(token, secret, options, callback)
      }
```

What would be the best way to make then token from `getToken` be available for `verify`?

Ethan-Arrowood commented:
Okay another big push. Fixed up the api, added synchronous  `decode`. Also fixed unit tests to bring code coverage back to 100%. Matteo said something about decorating the request and reply instances instead of fastify itself. I think this is a pretty easy fix if we move in that direction. Let me know your thoughts!

mcollina commented:
This current change does not make a lot of sense with the methods in the main `fastify` object`. I would actually __maintain the the previous functionality__, and add the new functionality via `decorateRequest` (verify) and `decorateReply` (sign/setHeader).

See https://github.com/fastify/fastify-auth/blob/master/example.js#L17-L49 on how it is currently used.

Ethan-Arrowood commented:
Okay, got it. I'll restore the original functionality and add the new functionality to the request or reply instances.

Ethan-Arrowood commented:
Major update again 😄 I've restored the original functions (pretty much copy and pasted back in). Then I added the `request.jwtVerify` and `reply.jwtSign` methods. I got the test coverage up to 98%. I think I'm only missing some basic logic branches -- I'll revisit tomorrow and try to get it to 100%.

mcollina commented:
the tests are failing on CI, can you have a look?

Ethan-Arrowood commented:
Oh the CI is meant to check for Node versions 4 - 8. It's failing on version 6 because I'm using async/await. Should I edit the code or just modify .travis.yml? Which versions of Node.js should this api support?

mcollina commented on package.json:
```diff
@@ -24,12 +24,16 @@
   },
   "homepage": "https://github.com/fastify/fastify-jwt#readme",
   "dependencies": {
+    "async": "^2.6.0",
```
can you remove async?

mcollina commented:
Node 4+.

Ethan-Arrowood commented:
If we need to support 4+, then do I need to modify the test file to not use ES6 stuff (async/await, parameter defaults)?

mcollina commented:
Yes exactly!

(We might argue to suppor only 6+, but it doesn't have those either)

Ethan-Arrowood commented:
Mwahaha supports Node 4+ now (isn't LTS for Node 4.x being removed when they release Node 10?)

mcollina commented on test.js:
```diff
@@ -1,95 +1,270 @@
+      uri: 'http://localhost:' + fastify.server.address().port + '/sign',
+      json: true
+    })
+      .then(function (sign) {
```
Can you inline those?

```js

rp({
  ...
}).then(function (sign) {
}).then(...)
````

mcollina commented on test.js:
```diff
@@ -1,95 +1,270 @@
+      if (err) { return reply.send(err) }
+      return reply.send({token})
+    })
+  })
```
can you make `jwtSign()` and `jwtVerify()` return a `Promise` if no callback is passed through?

This would enable a nice async/await syntax.

Ethan-Arrowood replied:
I hope I added this functionality correctly. I'll write some unit tests now to test it out.

Ethan-Arrowood commented:
Also, I was able to implement the `request.jwtVerify` method in my application with the RSA and JWKS things I wrote before 😄 its all coming together!

delvedor commented on README.md:
```diff
@@ -10,12 +10,13 @@ npm i fastify-jwt --save
-fastify.register(require('fastify-jwt'), { secret: 'supersecret' }, err => {
+fastify.register(require('fastify-jwt'), { 
+  secret: 'supersecret' 
+  }, err => {
```
Can you remove this callback? Is no longer supported :)

Ethan-Arrowood replied:
oops missed this

delvedor commented on README.md:
```diff
@@ -30,6 +31,121 @@ fastify.listen(3000, err => {
+
+fastify.get('/verify', function (request, reply) {
+  request.jwtVerify(function (err, decoded) {
+    if (err) { return reply.send(err) }
```
Here you can write `reply.send(err || decoded)`, same above :)

delvedor commented on jwt.js:
```diff
@@ -1,23 +1,36 @@
-function fastifyJWT (fastify, opts, next) {
-  if (!opts.secret) {
+function wrapStaticSecretInCallback (secret) {
+  return function (_, __, cb) {
```
Can you call these variable `request` and `reply`?

Ethan-Arrowood replied:
Renaming to request and payload in next commit

delvedor commented on jwt.js:
```diff
@@ -56,6 +69,79 @@ function fastifyJWT (fastify, opts, next) {
+    }
+    steed.waterfall([
+      function getSecret (callback) {
+        secretCallback(null, null, callback)
```
Why `null, null`?

Ethan-Arrowood replied:
The first null is replacing `request`. This method does not have access to the `request` object so I just left it as null. The second parameter is supposed to be payload and I have included that in the next commit.

delvedor replied:
Actually you can access the request object with `reply.request` (I know, it's weird :D)

Ethan-Arrowood commented:
🎉 woohoo

delvedor commented:
Landed in [`v0.2.0`](https://github.com/fastify/fastify-jwt/releases/tag/v0.2.0)!
